src/librustc/infer/resolve.rs

```diff
@@ -111,6 +111,43 @@ impl<'a, 'gcx, 'tcx> TypeVisitor<'tcx> for UnresolvedTypeFinder<'a, 'gcx, 'tcx>
     }
 }
 
+pub struct UnresolvedTypeCollector<'a, 'gcx: 'a+'tcx, 'tcx: 'a> {
+    infcx: &'a InferCtxt<'a, 'gcx, 'tcx>,
+    pub unresolved_types: Vec<(Span,Ty<'tcx>)>
+}
+
+impl<'a, 'gcx, 'tcx> UnresolvedTypeCollector<'a, 'gcx, 'tcx> {
+    pub fn new(infcx: &'a InferCtxt<'a, 'gcx, 'tcx>) -> Self {
+        UnresolvedTypeCollector { infcx, unresolved_types: vec![] }
+    }
+}
+
+impl<'a, 'gcx, 'tcx> TypeVisitor<'tcx> for UnresolvedTypeCollector<'a, 'gcx, 'tcx> {
+    fn visit_ty(&mut self, t: Ty<'tcx>) -> bool {
+        let t = self.infcx.shallow_resolve(t);
+        if t.has_infer_types() {
+            if let ty::Infer(ty::TyVar(ty_vid)) = t.sty {
+                let ty_vars = self.infcx.type_variables.borrow();
+                if let TypeVariableOrigin::TypeParameterDefinition(span, _name)
+                = *ty_vars.var_origin(ty_vid) {
+                    self.unresolved_types.push((span, t));
+                };
+
+                // Since we called `shallow_resolve` above, this must
+                // be an (as yet...) unresolved inference variable.
+                true
+            } else {
+                // Otherwise, visit its contents.
+                t.super_visit_with(self)
+            }
+        } else {
+            // Micro-optimize: no inference types at all Can't have unresolved type
```

Comment:
```suggestion
            // Micro-optimize: no inference types at all can't have unresolved type
```

Reply:
further nit: "all type variables in inference types must be resolved, no need to visit the contents"?